File: misc/image-to-text.js
const fs = require('fs');
const {createCanvas, loadImage} = require('canvas');

process.argv.slice(2).forEach(path => {
    loadImage(path).then(image => {
        const lines = [];

        const width = image.width;
        const height = image.height;

        const canvas = createCanvas(width, height);
        const context = canvas.getContext('2d');

        context.clearRect(0, 0, width, height);
        context.drawImage(image, 0, 0);
        let imageData = context.getImageData(0, 0, width, height);
        let data = imageData.data;

        for (let y = 0; y < height; y ++) {
            let split = [];

            for (let x = 0; x < width; x ++) {
                let i = ((y*width) + x) * 4;

                let s = '~';

                if (data[i+3] > 128) {
                    let r = data[i+0];
                    let g = data[i+1];
                    let b = data[i+2];

                    if (r == 85 && g == 85 && b == 85) {
                        s = 'a'; // use grayscale character #1
                    }
                    else if (r == 170 && g == 170 && b == 170) {
                        s = 'b'; // use grayscale character #2
                    }
                    else { // use one of 64 color characters
                        let mR = Math.floor(r/64);
                        let mG = Math.floor(g/64);
                        let mB = Math.floor(b/64);

                        let c = 0;

                        c += mR;
                        c = c << 2;
                        c += mG;
                        c = c << 2;
                        c += mB;

                        s = String.fromCharCode(c+33);
                    }
                }

                split.push(s);
            }

            const lineNumber = ('00'+y).substr(-2);

            lines.push([lineNumber+'d'].concat(split).join(' '));
        }

        const reversed = [];
        lines.forEach(line => {
            reversed.unshift(line.replace('d ', 'u '));
        });

        console.log(reversed.join('\n'));
        console.log(lines.join('\n'));
    });
});

File: image-to-text.py
import sys
import math
from PIL import Image

for path in sys.argv[1:]:
    image = Image.open(path).convert('RGBA')
    pixels = image.load()
    width = image.size[0]
    height = image.size[1]

    lines = []

    for y in range(height):
        split = ['%02dd' % y]

        for x in range(width):
            color = pixels[x, y] 

            s = '~'

            if len(color) < 4 or color[3] > 128:
                r = color[0]
                g = color[1]
                b = color[2]
               
                if r == 85 and g == 85 and b == 85:
                    s = 'a' # use grayscale character 1

                elif r == 170 and g == 170 and b == 170:
                    s = 'b' # use grayscale character 2

                else: # use one of 64 color characters
                    mR = math.floor(r/64)
                    mG = math.floor(g/64)
                    mB = math.floor(b/64)

                    c = 0;

                    c += mR
                    c = c << 2
                    c += mG
                    c = c << 2
                    c += mB

                    s = chr(c+33)

            split.append(s)

        lines.append(' '.join(split))

    reversed = []
    for line in lines:
        reversed.insert(0, (line.replace('d ', 'u ', 1)))

    print('\n'.join(reversed))
    print('\n'.join(lines))
